kernel/ipc.c

```diff
@@ -0,0 +1,40 @@
+#include "ipc.h"
+
+#include "com.h"
+#include "pagefault.h"
+#include "threads.h"
+#include "util.h"
+#include "memory-map.h"
+
+sendrecv_status sendrecv(sendrecv_op* usr_op) {
+  sendrecv_op op;
```

Comment:
I think it's clearer if you specifically copy just the send-message here, and specifically copy the recv-message back at the end.